usr/share/job/adapter.slurm.erb

```diff
@@ -24,13 +30,79 @@ register_control() {
 }
 
 <% if template.tag('script:type') == 'interactive' -%>
+<% if template.tag('session:host') == 'login' -%>
+#=========================
+#  Session orchestration
+#-------------------------
+# Functions used to orchestrate the creation of your interactive session on a
+# login node with X11 forwarding to a compute node.
+# XXX RESULTS_DIR might not be set yet.  Or could be wrong.
+SESSION_OUTPUT="${RESULTS_DIR}/session.output"
+
+on_login_host() {
+    [ -z $SLURM_JOBID ]
+}
+
+prepare_x11_forwarding() {
+    # Add our short hostname to the DISPLAY variable if we need to
+    if [ `echo $DISPLAY | cut -d: -f1 | egrep -c [a-z]` -lt 1 ] ; then
+        export DISPLAYNAME=`hostname -s`$DISPLAY
+    else
+        export DISPLAYNAME=$DISPLAY
+    fi
+}
+
+register_scheduler_id() {
+    stdbuf -oL sed -n 's/.*job \([0-9]*\) queued.*/\1/p' \
+        | head -n 1 \
+        | register_control "scheduler_id" 
+}
+
+request_allocation() {
+    # Request a slurm command-line session with any supplied variables, and
+    # run this job script on the allocation.
+    echo srun --pty "$@" "${BASH_SOURCE[0]}" $DISPLAYNAME
+    local submit_status
+    register_control "job_type" "BOOTSTRAPPING"
+    register_control "submit_status" "0"
+    srun --pty "$@" "${BASH_SOURCE[0]}" $DISPLAYNAME \
+        2> >( tee >( register_scheduler_id ) )
+    submit_status=$?
+    if [[ $submit_status -ne 0 ]] ; then
+        register_control "submit_status" $submit_status
+        register_control "job_type" "FAILED_SUBMISSION"
+        register_control "submit_stderr" "srun failed to submit job"
+    else
+        echo "Scheduler allocation has completed. Exiting..."
+        sleep 2
+    fi
+}
+
+configure_x11_forwarding() {
+    # Setup the originator to expect our X connection.
+    DISPLAYNAME=$1
+    DISPLAYHOST=`echo $DISPLAYNAME | cut -d: -f1`
+    this_hostname=`hostname -s`
+    setxhost="ssh $DISPLAYHOST 'export DISPLAY=$DISPLAYNAME && xhost $this_hostname'"
+    echo -ne "Enabling $DISPLAYHOST to accept our X-connection... "
+    eval $setxhost
+    export DISPLAY=$DISPLAYNAME
```

Comment:
Some of this output should be saved to results dir in addition to be sent to to the job scripts stdout/err.